### PR TITLE
James 1745 Handle attachments in GetMessages

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java [hunk 1]
```diff
@@ -148,15 +151,39 @@ private static void notImplemented(String input) {
...
+
+        Function<List<AttachmentId>, List<Attachment>> retrieveAttachments = retrieveAttachments(attachmentMapperFactory.getAttachmentMapper(mailboxSession));
+
         MailboxPath mailboxPath = value.getValue1();
-        return targetStream.map(x -> Pair.with(x, mailboxPath));
+        return targetStream.map(Throwing
+                .function((MailboxMessage message) -> new JamesMessage(message, retrieveAttachments.apply(message.getAttachmentsIds()), mailboxPath)));
+    }
+
+    private class JamesMessage {
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: The name tell me nothing : everything is james ...

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java [hunk 2]
```diff
@@ -126,17 +134,12 @@ private static void notImplemented(String input) {
...
 
     
-    private Function<Pair<MailboxMessage, MailboxPath>, Message> toJmapMessage(MailboxSession mailboxSession) {
-        return (value) -> {
-            MailboxMessage messageResult = value.getValue0();
-            MailboxPath mailboxPath = value.getValue1();
-            return Message.fromMailboxMessage(messageResult, uid -> new MessageId(mailboxSession.getUser(), mailboxPath , uid));
-        };
+    private Function<JamesMessage, Message> toJmapMessage(MailboxSession mailboxSession) {
+        return (jamesMessage) -> Message.fromMailboxMessage(jamesMessage.mailboxMessage, jamesMessage.attachments,
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: you could do better with indentation

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java [hunk 3]
```diff
@@ -148,15 +151,39 @@ private static void notImplemented(String input) {
...
                              mailboxPath
                              );
-         })
-                .andThen(this::iteratorToStream);
+                })
+                .andThen(Throwing.function((pair) -> iteratorToStream(pair, mailboxSession)));
     }
     
-    private Stream<Pair<MailboxMessage, MailboxPath>> iteratorToStream(Pair<Iterator<MailboxMessage>, MailboxPath> value) {
+    private Stream<JamesMessage> iteratorToStream(Pair<Iterator<MailboxMessage>, MailboxPath> value, MailboxSession mailboxSession) throws MailboxException {
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: this method no longer do what its name says

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java [hunk 4]
```diff
@@ -148,15 +151,39 @@ private static void notImplemented(String input) {
...
         Iterable<MailboxMessage> iterable = () -> value.getValue0();
         Stream<MailboxMessage> targetStream = StreamSupport.stream(iterable.spliterator(), false);
-        
+
+        Function<List<AttachmentId>, List<Attachment>> retrieveAttachments = retrieveAttachments(attachmentMapperFactory.getAttachmentMapper(mailboxSession));
+
         MailboxPath mailboxPath = value.getValue1();
-        return targetStream.map(x -> Pair.with(x, mailboxPath));
+        return targetStream.map(Throwing
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: I think you do to much on a single map step

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java (hunk 2 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: OK

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java (hunk 1 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: OK

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java (hunk 3 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: OK

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java [hunk 5]
```diff
@@ -185,7 +189,8 @@ public void getMessagesShouldReturnMessagesWhenAvailable() throws Exception {
             .body(ARGUMENTS + ".list[0].isUnread", equalTo(true))
             .body(ARGUMENTS + ".list[0].preview", equalTo("testmail"))
             .body(ARGUMENTS + ".list[0].headers", equalTo(ImmutableMap.of("subject", "my test subject")))
-            .body(ARGUMENTS + ".list[0].date", equalTo("2014-10-30T14:12:00Z"));
+            .body(ARGUMENTS + ".list[0].date", equalTo("2014-10-30T14:12:00Z"))
+            .body(ARGUMENTS + ".list[0].attachments", empty());
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: ATTACHMENTS

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java [hunk 6]
```diff
@@ -331,4 +336,34 @@ public void getMessagesShouldReturnMandatoryPropertiesMessagesWhenNotAsked() thr
...
+            .statusCode(200)
+            .body(NAME, equalTo("messages"))
+            .body(ARGUMENTS + ".list", hasSize(1))
+            .body(ATTACHMENTS, hasSize(2))
+            .body(FIRST_ATTACHMENT + ".blobId", equalTo("223a76c0e8c1b1762487d8e0598bd88497d73ef2"))
+            .body(FIRST_ATTACHMENT + ".type", equalTo("image/jpeg"))
+            .body(FIRST_ATTACHMENT + ".size", equalTo(846))
+            .body(SECOND_ATTACHMENT + ".blobId", equalTo("58aa22c2ec5770fb9e574ba19008dbfc647eba43"))
+            .body(SECOND_ATTACHMENT + ".type", equalTo("image/jpeg"))
+            .body(SECOND_ATTACHMENT + ".size", equalTo(597));
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: Can you check hasAttachment too?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java [hunk 7]
```diff
@@ -181,6 +180,20 @@ private static boolean isHtml(IndexableMessage im) {
             && im.getSubType().equals(MediaType.HTML_UTF_8.subtype());
     }
 
+    private static List<Attachment> getAttachments(List<org.apache.james.mailbox.store.mail.model.Attachment> attachments) {
+        return attachments.stream()
+                .map(Message::fromAttachment)
+                .collect(org.apache.james.util.streams.Collectors.toImmutableList());
+    }
+
+    private static Attachment fromAttachment(org.apache.james.mailbox.store.mail.model.Attachment attachment) {
```
Comment (rouazana) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: fromMailboxAttachment or fromStoreAttachment?

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/AttachmentTest.java [hunk 8]
```diff
@@ -56,26 +56,20 @@ public void buildShouldThrowWhenTypeIsEmpty() {
         Attachment.builder().blobId("blobId").type("").name("name").size(123).build();
     }
     
-    @Test(expected=IllegalStateException.class)
-    public void buildShouldThrowWhenNameIsEmpty() {
-        Attachment.builder().blobId("blobId").type("type").name("").size(123).build();
```
Comment (rouazana) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/AttachmentTest.java]: Do we accept an empty string as name?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java (hunk 4 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: Let me known what you think of that...

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java [hunk 9]
```diff
@@ -152,34 +154,84 @@ private static void notImplemented(String input) {
...
+        }
+
+        public static class Builder {
+
+            private MailboxMessage mailboxMessage;
+            private List<AttachmentId> attachmentIds;
+            private MailboxPath mailboxPath;
+            private Function<List<AttachmentId>, List<Attachment>> retrieveAttachments;
+
+            Builder() {
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: private

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java [hunk 10]
```diff
@@ -152,34 +154,84 @@ private static void notImplemented(String input) {
...
         Iterable<MailboxMessage> iterable = () -> value.getValue0();
         Stream<MailboxMessage> targetStream = StreamSupport.stream(iterable.spliterator(), false);
 
         Function<List<AttachmentId>, List<Attachment>> retrieveAttachments = retrieveAttachments(attachmentMapperFactory.getAttachmentMapper(mailboxSession));
 
         MailboxPath mailboxPath = value.getValue1();
-        return targetStream.map(Throwing
-                .function((MailboxMessage message) -> new JamesMessage(message, retrieveAttachments.apply(message.getAttachmentsIds()), mailboxPath)));
+        return targetStream
+                .map(message -> CompletedMailboxMessage.builder().mailboxMessage(message).attachmentIds(message.getAttachmentsIds()))
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: looks ok

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java (hunk 5 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: OK

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java (hunk 7 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: OK

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java (hunk 9 above)
Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: OK

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/model/AttachmentTest.java (hunk 8 above)
Comment (aduprat) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/model/AttachmentTest.java]: This field is not yet supported.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java (hunk 6 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessagesMethodTest.java]: OK

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java [hunk 11]
```diff
@@ -181,6 +180,20 @@ private static boolean isHtml(IndexableMessage im) {
             && im.getSubType().equals(MediaType.HTML_UTF_8.subtype());
     }
 
+    private static List<Attachment> getAttachments(List<org.apache.james.mailbox.store.mail.model.Attachment> attachments) {
```
Comment (chibenwa) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: FQDN ? Where is the conflict ?

**Comment by rouazana** [comment]
:+1:

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java [hunk 12]
```diff
@@ -87,7 +88,8 @@ public static Message fromMailboxMessage(MailboxMessage mailboxMessage,
                 .preview(getPreview(im))
                 .textBody(getTextBody(im))
                 .htmlBody(getHtmlBody(im))
-                .attachments(getAttachments(attachments))
+                .attachments(asJmapAttachments)
+                .hasAttachment(!asJmapAttachments.isEmpty())
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: sorry but I don't agree : if you have attachments, you should not have to put a boolean in a hasAttachment in a field, it's just another view of the same information.

Comment (aduprat) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/model/Message.java]: OK

**Comment by rouazana** [comment]
:+1:

**Comment by mbaechler** [comment]
👍

**Pull request merged**